### PR TITLE
Compatibility with gmpxx

### LINKED ISSUE #145: 1.3.2 fails to build: Coordinate.cpp:270:72: error: no matching function for call to 'round(__gmp_expr<__mpq_struct [1], __gmp_binary_expr<__gmp_expr<__mpq_struct [1], __mpq_struct [1]>, long int, __gmp_binary_multiplies> >)'
SFCGAL 1.3.2 fails to build with CGAL 4.11:
```
[  0%] Building CXX object src/CMakeFiles/SFCGAL.dir/Coordinate.cpp.o
cd /build/sfcgal-1.3.2/obj-x86_64-linux-gnu/src && /usr/bin/c++  -DBOOST_ALL_DYN_LINK -DBOOST_ALL_NO_LIB -DSFCGAL_BUILD_SHARED -DSFCGAL_EXPORTS -I/build/sfcgal-1.3.2/obj-x86_64-linux-gnu/include -isystem /usr/include/x86_64-linux-gnu  -g -O2 -fdebug-prefix-map=/build/sfcgal-1.3.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -Wall -Wextra -Wno-long-long -pedantic -Wpointer-arith -Wcast-align -Wcast-qual -Wno-overloaded-virtual -Wformat=2 -Winit-self -Wmissing-include-dirs -Wwrite-strings -Wno-error=undef -O2 -g -DNDEBUG -fPIC   -o CMakeFiles/SFCGAL.dir/Coordinate.cpp.o -c /build/sfcgal-1.3.2/src/Coordinate.cpp
[  1%] Building CXX object src/CMakeFiles/SFCGAL.dir/Envelope.cpp.o
[  1%] Building CXX object src/CMakeFiles/SFCGAL.dir/Exception.cpp.o
cd /build/sfcgal-1.3.2/obj-x86_64-linux-gnu/src && /usr/bin/c++  -DBOOST_ALL_DYN_LINK -DBOOST_ALL_NO_LIB -DSFCGAL_BUILD_SHARED -DSFCGAL_EXPORTS -I/build/sfcgal-1.3.2/obj-x86_64-linux-gnu/include -isystem /usr/include/x86_64-linux-gnu  -g -O2 -fdebug-prefix-map=/build/sfcgal-1.3.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -Wall -Wextra -Wno-long-long -pedantic -Wpointer-arith -Wcast-align -Wcast-qual -Wno-overloaded-virtual -Wformat=2 -Winit-self -Wmissing-include-dirs -Wwrite-strings -Wno-error=undef -O2 -g -DNDEBUG -fPIC   -o CMakeFiles/SFCGAL.dir/Envelope.cpp.o -c /build/sfcgal-1.3.2/src/Envelope.cpp
cd /build/sfcgal-1.3.2/obj-x86_64-linux-gnu/src && /usr/bin/c++  -DBOOST_ALL_DYN_LINK -DBOOST_ALL_NO_LIB -DSFCGAL_BUILD_SHARED -DSFCGAL_EXPORTS -I/build/sfcgal-1.3.2/obj-x86_64-linux-gnu/include -isystem /usr/include/x86_64-linux-gnu  -g -O2 -fdebug-prefix-map=/build/sfcgal-1.3.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -Wall -Wextra -Wno-long-long -pedantic -Wpointer-arith -Wcast-align -Wcast-qual -Wno-overloaded-virtual -Wformat=2 -Winit-self -Wmissing-include-dirs -Wwrite-strings -Wno-error=undef -O2 -g -DNDEBUG -fPIC   -o CMakeFiles/SFCGAL.dir/Exception.cpp.o -c /build/sfcgal-1.3.2/src/Exception.cpp
[  2%] Building CXX object src/CMakeFiles/SFCGAL.dir/Geometry.cpp.o
cd /build/sfcgal-1.3.2/obj-x86_64-linux-gnu/src && /usr/bin/c++  -DBOOST_ALL_DYN_LINK -DBOOST_ALL_NO_LIB -DSFCGAL_BUILD_SHARED -DSFCGAL_EXPORTS -I/build/sfcgal-1.3.2/obj-x86_64-linux-gnu/include -isystem /usr/include/x86_64-linux-gnu  -g -O2 -fdebug-prefix-map=/build/sfcgal-1.3.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -Wall -Wextra -Wno-long-long -pedantic -Wpointer-arith -Wcast-align -Wcast-qual -Wno-overloaded-virtual -Wformat=2 -Winit-self -Wmissing-include-dirs -Wwrite-strings -Wno-error=undef -O2 -g -DNDEBUG -fPIC   -o CMakeFiles/SFCGAL.dir/Geometry.cpp.o -c /build/sfcgal-1.3.2/src/Geometry.cpp
/build/sfcgal-1.3.2/src/Coordinate.cpp: In member function 'CGAL::Lazy_kernel_generic_base<CGAL::Simple_cartesian<__gmp_expr<__mpq_struct [1], __mpq_struct [1]> >, CGAL::Simple_cartesian<CGAL::Interval_nt<false> >, CGAL::Cartesian_converter<CGAL::Simple_cartesian<__gmp_expr<__mpq_struct [1], __mpq_struct [1]> >, CGAL::Simple_cartesian<CGAL::Interval_nt<false> > >, CGAL::Epeck>::FT SFCGAL::RoundVisitor::_roundFT(const FT&) const':
/build/sfcgal-1.3.2/src/Coordinate.cpp:270:72: error: no matching function for call to 'round(__gmp_expr<__mpq_struct [1], __gmp_binary_expr<__gmp_expr<__mpq_struct [1], __mpq_struct [1]>, long int, __gmp_binary_multiplies> >)'
                                SFCGAL::round( v.exact() * _scaleFactor ),
                                                                        ^
```

This issue was also reported by Joachim Reichel in [Debian Bug #876521](https://bugs.debian.org/876521)

### PR DESCRIPTION
**Pull request by sloriot**
Fixes #145

### CODE REVIEW

**Comment by mhugo** [comment]
@sloriot Thanks. Do you know if this passes tests ?

**Comment by sebastic** [comment]
Applying 5c495ce98c404f891eb0e2c92c2d73b94af3ef23 and the changes from this PR on top of 1.3.2 still fails to build with CGAL 4.11.1 in Debian unstable:
```
/usr/bin/c++  -g -O2 -fdebug-prefix-map=/build/sfcgal-1.3.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -Wall -Wextra -Wno-long-long -pedantic -Wpointer-arith -Wcast-align -Wcast-qual -Wno-overloaded-virtual -Wformat=2 -Win
it-self -Wmissing-include-dirs -Wwrite-strings -Wno-error=undef -O2 -g -DNDEBUG  -Wl,-z,relro -Wl,-z,now -rdynamic CMakeFiles/test-regress-polygon_triangulator.dir/main.cpp.o  -o test-regress-polygon_triangulator -Wl,-rpath,/build/sfcgal-1.3.2/obj-x86_64-linux-gnu/src: .
./../../src/libSFCGAL.so.1.3.2 -lboost_program_options -lboost_chrono -lboost_filesystem -lboost_timer -lboost_unit_test_framework -lboost_program_options -lboost_date_time -lboost_thread -lboost_system -lboost_serialization -lboost_atomic -lpthread -lCGAL -lCGAL_Core -l
mpfr -lgmp -lboost_program_options -lboost_chrono -lboost_filesystem -lboost_timer -lboost_unit_test_framework -lboost_program_options -lboost_chrono -lboost_filesystem -lboost_timer -lboost_unit_test_framework -lboost_date_time -lboost_thread -lboost_system -lboost_seri
alization -lboost_atomic -lpthread 
../../../src/libSFCGAL.so.1.3.2: undefined reference to `operator<<(std::ostream&, __mpq_struct const*)'
../../../src/libSFCGAL.so.1.3.2: undefined reference to `operator<<(std::ostream&, __mpz_struct const*)'
../../../src/libSFCGAL.so.1.3.2: undefined reference to `SFCGAL::round(__gmp_expr<__mpq_struct [1], __mpq_struct [1]> const&)'
../../../src/libSFCGAL.so.1.3.2: undefined reference to `operator>>(std::istream&, __mpz_struct*)'
../../../src/libSFCGAL.so.1.3.2: undefined reference to `operator>>(std::istream&, __mpq_struct*)'
collect2: error: ld returned 1 exit status
make[3]: *** [test/regress/polygon_triangulator/CMakeFiles/test-regress-polygon_triangulator.dir/build.make:131: test/regress/polygon_triangulator/test-regress-polygon_triangulator] Error 1
```
```
cd /build/sfcgal-1.3.2/obj-x86_64-linux-gnu/test/regress/standalone && /usr/bin/cmake -E cmake_link_script CMakeFiles/standalone-regress-test-SFCGAL.dir/link.txt --verbose=1
/usr/bin/c++  -g -O2 -fdebug-prefix-map=/build/sfcgal-1.3.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -Wall -Wextra -Wno-long-long -pedantic -Wpointer-arith -Wcast-align -Wcast-qual -Wno-overloaded-virtual -Wformat=2 -Win
it-self -Wmissing-include-dirs -Wwrite-strings -Wno-error=undef -O2 -g -DNDEBUG  -Wl,-z,relro -Wl,-z,now -rdynamic CMakeFiles/standalone-regress-test-SFCGAL.dir/SFCGAL/AreaTest.cpp.o CMakeFiles/standalone-regress-test-SFCGAL.dir/SFCGAL/ConstraintDelaunayTriangulationTest
.cpp.o CMakeFiles/standalone-regress-test-SFCGAL.dir/SFCGAL/DistanceTest.cpp.o CMakeFiles/standalone-regress-test-SFCGAL.dir/SFCGAL/IntersectsTest.cpp.o CMakeFiles/standalone-regress-test-SFCGAL.dir/SFCGAL/StraightSkeletonTest.cpp.o CMakeFiles/standalone-regress-test-SFC
GAL.dir/SFCGAL/TriangulatePolygonTest.cpp.o CMakeFiles/standalone-regress-test-SFCGAL.dir/SFCGAL/WaveFrontObjTest.cpp.o CMakeFiles/standalone-regress-test-SFCGAL.dir/SFCGAL/WktTest.cpp.o CMakeFiles/standalone-regress-test-SFCGAL.dir/TestModule.cpp.o  -o standalone-regres
s-test-SFCGAL -Wl,-rpath,/build/sfcgal-1.3.2/obj-x86_64-linux-gnu/src: ../../../src/libSFCGAL.so.1.3.2 -lboost_program_options -lboost_chrono -lboost_filesystem -lboost_timer -lboost_unit_test_framework -lboost_program_options -lboost_date_time -lboost_thread -lboost_sys
tem -lboost_serialization -lboost_atomic -lpthread -lCGAL -lCGAL_Core -lmpfr -lgmp -lboost_program_options -lboost_chrono -lboost_filesystem -lboost_timer -lboost_unit_test_framework -lboost_program_options -lboost_chrono -lboost_filesystem -lboost_timer -lboost_unit_tes
t_framework -lboost_date_time -lboost_thread -lboost_system -lboost_serialization -lboost_atomic -lpthread 
../../../src/libSFCGAL.so.1.3.2: undefined reference to `operator<<(std::ostream&, __mpq_struct const*)'
../../../src/libSFCGAL.so.1.3.2: undefined reference to `operator<<(std::ostream&, __mpz_struct const*)'
../../../src/libSFCGAL.so.1.3.2: undefined reference to `SFCGAL::round(__gmp_expr<__mpq_struct [1], __mpq_struct [1]> const&)'
../../../src/libSFCGAL.so.1.3.2: undefined reference to `operator>>(std::istream&, __mpz_struct*)'
../../../src/libSFCGAL.so.1.3.2: undefined reference to `operator>>(std::istream&, __mpq_struct*)'
collect2: error: ld returned 1 exit status
make[3]: *** [test/regress/standalone/CMakeFiles/standalone-regress-test-SFCGAL.dir/build.make:339: test/regress/standalone/standalone-regress-test-SFCGAL] Error 1
```
```
/build/sfcgal-1.3.2/test/unit/SFCGAL/KernelTest.cpp:67:29: error: no matching function for call to 'CGAL::Lazy_exact_nt<__gmp_expr<__mpq_struct [1], __mpq_struct [1]> >::Lazy_exact_nt(std::__cxx11::basic_ostringstream<char>::__string_type)'
     Kernel::FT b( oss.str() );
                             ^
In file included from /usr/include/CGAL/internal/Exact_type_selector.h:33:0,
                 from /usr/include/CGAL/Filtered_kernel.h:35,
                 from /usr/include/CGAL/Exact_predicates_exact_constructions_kernel.h:29,
                 from /build/sfcgal-1.3.2/obj-x86_64-linux-gnu/include/SFCGAL/Kernel.h:24,
                 from /build/sfcgal-1.3.2/test/unit/SFCGAL/KernelTest.cpp:24:
```
```
/usr/include/CGAL/Lazy_exact_nt.h:371:3: error: no type named 'type' in 'struct boost::enable_if<boost::mpl::and_<boost::mpl::or_<boost::is_arithmetic<std::__cxx11::basic_string<char> >, std::is_enum<std::__cxx11::basic_string<char> >, mpl_::bool_<false>, mpl_::bool_<false>, mpl_::bool_<false> >, boost::mpl::not_<boost::is_same<std::__cxx11::basic_string<char>, __gmp_expr<__mpq_struct [1], __mpq_struct [1]> > >, mpl_::bool_<true>, mpl_::bool_<true>, mpl_::bool_<true> >, void*>'
/usr/include/CGAL/Lazy_exact_nt.h:366:3: note: candidate: CGAL::Lazy_exact_nt<ET_>::Lazy_exact_nt(CGAL::Lazy_exact_nt<ET_>::Self_rep*) [with ET_ = __gmp_expr<__mpq_struct [1], __mpq_struct [1]>; CGAL::Lazy_exact_nt<ET_>::Self_rep = CGAL::Lazy_rep<CGAL::Interval_nt<false>, __gmp_expr<__mpq_struct [1], __mpq_struct [1]>, CGAL::To_interval<__gmp_expr<__mpq_struct [1], __mpq_struct [1]> > >]
   Lazy_exact_nt (Self_rep *r)
   ^~~~~~~~~~~~~
```

**Comment by sloriot** [comment]
@mhugo I haven't tried.
@sebastic I'm surprized, did you use cmake to configure?

**Comment by sebastic** [comment]
> I'm surprized, did you use cmake to configure?

Yes, it's the only buildsystem supported by SFCAL.

Full buildlog: [sfcgal_1.3.2-1_amd64.txt](https://github.com/Oslandia/SFCGAL/files/1815755/sfcgal_1.3.2-1_amd64.txt)

**Comment by sloriot** [comment]
I probably did not built everything, let me try again.

**Comment by sloriot** [comment]
`-DSFCGAL_BUILD_TESTS=ON` was needed to reproduce the error.
It could reproduce and fix the errors. Please give it another try.
@mhugo I don't know if you'll like the cmake modifications.

**Comment by sebastic** [comment]
The CMake modifications break the build:
```
/usr/bin/ld: cannot find -lCGAL::CGAL
/usr/bin/ld: cannot find -lCGAL::CGAL_Core
```

**Comment by sloriot** [comment]
what is you cmake version?

**Comment by sebastic** [comment]
```
$ grep "cmake (" sfcgal_1.3.2-1_amd64.txt
Unpacking cmake (3.10.2-1) ...
Setting up cmake (3.10.2-1) ...
```

**Comment by sloriot** [comment]
The issue is in the `cmake/Modules/FindCGAL.cmake` that is not working correctly in my opinion. I did not understand why it was not compiling for me, so I simply removed it.
Asking for `CGAL_DIR` to be the root of the include dir is not a good idea because this breaks the _Config_ mode of `find_package` of cmake [[1](https://cmake.org/cmake/help/v3.0/command/find_package.html)]. @sebastic if you remove that file and set `CGAL_DIR` to point to the directory containing `CGALConfig.cmake` (`/usr/lib/x86_64-linux-gnu/cmake/CGAL` for the default package in my OS) I think it should build correctly.

**Comment by sebastic** [comment]
Removing `FindCGAL.cmake` is sufficient let the build succeeed, as `/usr/lib/x86_64/cmake` is already part of the `CMAKE_MODULE_PATH`. 

Unfortunately the tests fail:
```
33% tests passed, 2 tests failed out of 3

Total Test time (real) =  14.70 sec

The following tests FAILED:
          1 - unit-test (Failed)
          2 - standalone-regress-test (Failed)
```
Buildlog: [sfcgal_1.3.2-1_amd64.txt](https://github.com/Oslandia/SFCGAL/files/1816753/sfcgal_1.3.2-1_amd64.txt)

**Comment by sebastic** [comment]
@sloriot, can you have a look at fixing the tests?

**Comment by sloriot** [comment]
I'm rather busy for the moment, I'll have a look as soon as I can.

**Comment by sloriot** [comment]
Should be fixed now.
@mhugo What shall we do with the `CMakeLists.txt`?

File: src/detail/tools/InputStreamReader.h [hunk 1]
```diff
@@ -150,7 +150,7 @@ class BasicInputStreamReader {
             skipWhiteSpaces();
         }
 
-        if ( _s >> value ) {
+        if ( CGAL::Input_rep<T>(value)(_s) ) {
```
Comment (lrineau) [inline src/detail/tools/InputStreamReader.h]:
```c++
if(_s >> CGAL::iformat(value))
```

**Comment by sebastic** [comment]
With or without @lrineau's suggestion, the unit-test still fails, 8de5943 does fix the standalone-regress-test.

Buildlog: [sfcgal_1.3.2-1_amd64.txt](https://github.com/Oslandia/SFCGAL/files/1839232/sfcgal_1.3.2-1_amd64.txt)

File: test/garden/CMakeLists.txt [hunk 2]
```diff
@@ -6,12 +6,10 @@ add_executable( ${REGRESS_NAME} ${SFCGAL_REGRESS_GARDEN_TEST_SOURCES} )
 
 target_link_libraries( ${REGRESS_NAME}
 	SFCGAL
-	${Boost_LIBRARIES}
-	${CGAL_LIBRARIES}
+        CGAL::CGAL
+        CGAL::CGAL_Core
```
Comment (mhugo) [inline test/garden/CMakeLists.txt]: @sloriot I don't know this syntax. What differences with the previous situation ? (i.e. what do you fix here ?)

File: test/garden/CMakeLists.txt [hunk 3]
```diff
@@ -6,12 +6,10 @@ add_executable( ${REGRESS_NAME} ${SFCGAL_REGRESS_GARDEN_TEST_SOURCES} )
...
 )
-if( ${SFCGAL_WITH_MPFR} )
-  target_link_libraries( ${REGRESS_NAME} ${MPFR_LIBRARIES} )
-endif( ${SFCGAL_WITH_MPFR} )
+target_link_libraries( ${REGRESS_NAME} ${CGAL_3RD_PARTY_LIBRARIES} )
```
Comment (mhugo) [inline test/garden/CMakeLists.txt]: @sloriot so it seems to reduce 3 lines to 1. I'm happy with that :)

File: test/garden/CMakeLists.txt (hunk 2 above)
Comment (lrineau) [reply test/garden/CMakeLists.txt]: That syntax is a new one, using imported targets. That syntax exists since CGAL-4.7, but it has been powered-up with CGAL-4.12-beta1, because now the imported targets know about their own dependencies: if you link with `CGAL::CGAL`, then you get automatically:
  - the include path of CGAL,
  - the include path of Boost, GMP, MPFR...,
  - link with CGAL,
  - link with with Boost Thread (if needed), and with GMP, and MPFR.

File: test/garden/CMakeLists.txt (hunk 3 above)
Comment (lrineau) [reply test/garden/CMakeLists.txt]: It could even be one line:
```cmake
target_link_libraries( ${REGRESS_NAME} SFCGAL CGAL::CGAL_Core ${CGAL_3RD_PARTY_LIBRARIES} )
```

Comment (lrineau) [reply test/garden/CMakeLists.txt]: And, actually, I think the variable `CGAL_3RD_PARTY_LIBRARIES` is probably empty, since CGAL-4.12-beta1.

**Comment by sloriot** [comment]
@mhugo Another issue is shall we remove the FindCGAL.cmake shipped with SFCGAL (since it prevents from building SFCGAL with my setting and @sebastic apparently)?

**Comment by mhugo** [comment]
@sloriot I guess it is ok to remove FindCGAL.cmake

**Comment by sloriot** [comment]
Travis seems to be happy now with my latest changes. @mhugo is 174dbb7 an issue for you?

**Comment by sebastic** [comment]
The Debian package build also succeeds with 174dbb7.

**Comment by vmora** [comment]
I compiled successfully and passed the tests, so I merged.

@mhugo shall I release ?

**Comment by mhugo** [comment]
@vmora yes please

**Comment by vmora** [comment]
@mhugo done
